balancing/balance_breaker.go

```diff
@@ -0,0 +1,615 @@
+func (counter *lengthDelimitedCounter) Percentile(percentile float64) float64 {
+	snapshot := make([]float64, len(counter.values))
+	copy(snapshot, counter.values)
+	sort.Float64s(snapshot)
+	pertcentileIndex := int(math.Floor(float64(len(snapshot)) * percentile))
+	return snapshot[pertcentileIndex]
+}
+
+func (counter *lengthDelimitedCounter) Reset() {
+	for idx := range counter.values {
+		counter.values[idx] = 0
+	}
+}
+
+type breakerState int
+
+const (
+	open     breakerState = 0
+	halfopen              = iota
+	closed                = iota
+)
+
+func newOpenStateTracker(start time.Time, changeDelay, maxDelay time.Duration) *openStateTracker {
+	return &openStateTracker{
+		lastChange:  start,
+		state:       open,
+		changeDelay: changeDelay,
+		maxDelay:    maxDelay,
+	}
+}
+
+type openStateTracker struct {
+	state          breakerState
+	lastChange     time.Time
+	changeDelay    time.Duration
+	maxDelay       time.Duration
+	closeIteration float64
+}
+
+func (tracker *openStateTracker) currentDelay() time.Duration {
+	multiplier := int(math.Pow(2, tracker.closeIteration))
+	delayDuration := tracker.changeDelay * time.Duration(multiplier)
+
+	if delayDuration < tracker.maxDelay {
+		return delayDuration
+	}
+
+	return tracker.maxDelay
+}
+
+func (tracker *openStateTracker) currentState(now time.Time, limitsExceeded bool) (breakerState, bool) {
+
+	if limitsExceeded && tracker.state != open {
+		tracker.state = open
+		tracker.lastChange = now
+		tracker.closeIteration++
+		return tracker.state, true
+	}
+
+	changed := false
+	if now.Sub(tracker.lastChange) < tracker.currentDelay() {
+		return tracker.state, changed
+	}
+
+	changed = true
+	tracker.lastChange = now
+	if tracker.state == open {
+		tracker.state = halfopen
+		return halfopen, changed
+	}
+
+	if tracker.state == halfopen {
+		if limitsExceeded {
+			tracker.state = open
+			tracker.closeIteration++
+		} else {
+			tracker.state = closed
+		}
+	}
+	return tracker.state, changed
+}
+
+// MeasuredStorage coordinates metrics collection
+type MeasuredStorage struct {
+	Node
+	Breaker
+	http.RoundTripper
+	Name           string
+	watcherStarted bool
+}
+
+// RoundTrip implements http.RoundTripper
+func (ms *MeasuredStorage) RoundTrip(req *http.Request) (*http.Response, error) {
+	start := time.Now()
+	reqID, _ := req.Context().Value(log.ContextreqIDKey).(string)
+	log.Debugf("MeasuredStorage %s: Got request id %s\n", ms.Name, reqID)
+	resp, err := ms.RoundTripper.RoundTrip(req)
+	duration := time.Since(start)
+	success := backend.IsSuccessful(resp, err)
+	open := ms.Breaker.Record(duration, success)
+	log.Debugf("MeasuredStorage %s: Request %s took %s was successful: %t, opened breaker %t\n", ms.Name, reqID, duration, success, open)
+
+	ms.Node.UpdateTimeSpent(duration)
+	ms.Node.SetActive(!open)
+	raportMetrics(ms.RoundTripper, start, open)
+	return resp, err
+}
+
+// IsActive checks Breaker status propagates it to Node compound
+func (ms *MeasuredStorage) IsActive() bool {
```

Comment:
i don't like the fact that the method's name hides what it's actually doing and you have to take a look at the doc to know the whole truth.